pennylane_lightning/core/lightning_interpreter.py

```diff
@@ -0,0 +1,182 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains a class for executing plxpr using default qubit tools.
+"""
+from copy import copy
+
+import jax
+from pennylane.capture import disable, enable
+from pennylane.capture.base_interpreter import PlxprInterpreter
+from pennylane.capture.primitives import (
+    adjoint_transform_prim,
+    cond_prim,
+    ctrl_transform_prim,
+    for_loop_prim,
+    measure_prim,
+    while_loop_prim,
+)
+from pennylane.measurements import MidMeasureMP, Shots
+
+from ._measurements_base import LightningBaseMeasurements
+from ._state_vector_base import LightningBaseStateVector
+
+
+class LightningInterpreter(PlxprInterpreter):
+    """A class that can interpret pennylane variant JAXPR.
+
+    Args:
+        state (LightningBaseStateVector): the class containing the statevector
+        measurement_class (type[LightningBaseMeasurement]): The type to use to perform
+            measurements on the statevector
+        shots (Shots): the number of shots to use. Shot vectors are not yet supported.
+
+    .. code-block:: python
+
+        import pennylane as qml
+        import jax
+        qml.capture.enable()
+
+        statevector = LightningStateVector(num_wires=3, dtype=np.complex128)
+        measurement_class = LightningMeasurements
+
+        interpreter = LightningInterpreter(statevector, measurement_class, shots=Shots(None))
+
+        def f(x):
+            @qml.for_loop(3)
+            def loop(i, y):
+                qml.RX(y, i)
+                return y + 0.5
+            loop(x)
+            return [qml.expval(qml.Z(i)) for i in range(3)]
+
+        jaxpr = jax.make_jaxpr(f)(1.2)
+
+        interpreter(f)(0.0), interpreter.eval(jaxpr.jaxpr, jaxpr.consts, 0.0)
+
+    .. code-block::
+
+        ([1.0, 0.8775825618903728, 0.5403023058681395], [1.0, 0.8775825618903728, 0.5403023058681395])
+    """
+
+    def __init__(
+        self,
+        state: LightningBaseStateVector,
+        measurement_class: type[LightningBaseMeasurements],
+        shots: Shots = Shots(None),
+    ):
+        self.state = state
+        self.measurement_class = measurement_class
+        self.shots = shots
+        if self.shots.has_partitioned_shots:
+            raise NotImplementedError("LightningInterpreter does not support partitioned shots.")
+        self.reset = True
+        super().__init__()
+
+    def setup(self) -> None:
+        """Reset the state if necessary."""
+        if self.reset:
+            self.state.reset_state()
+            self.reset = False  # copies will have reset=False and wont reset state
+
+    def cleanup(self) -> None:
+        """Indicate that the state will need to be reset if this instance is reused."""
+        self.reset = True
+
+    def interpret_operation(self, op):
```

Comment:
Are there any concerns with implementing `interpret_operations` in favour of avoiding these `interpret_operation` and `[op]` calls for each operation?

Reply:
So how this works is by iterating through the jaxpr equations, and calling this `interpret_operation` method any time we hit an operation.

We theoretically could serialize the entire jaxpr into a list of operations, then apply all the operations together, but I don't really see the need. Especially since all the `apply_operations` method just use a normal for loop internally.